fix: drop getters and setters when diffing objects for error

SimenB opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

Fixes #9745. Not sure if it makes sense to drop the fix from #9575? I can check if it's `writeable` at the same time I now check for a setter, which makes the test added at that time pass.

/cc @WeiAnAn 

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Test plan

Unit test added.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

SimenB commented on CHANGELOG.md:
```diff
@@ -3,6 +3,8 @@
 ### Features
 
 ### Fixes
+
```
CI is failing about this missing newline, snuck it in here

SimenB commented on packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts:
```diff
@@ -59,7 +59,9 @@ function deepCyclicCopyObject<T>(object: T, cycles: WeakMap<any, any>): T {
 
-    if (!('set' in descriptor)) {
+    if ('set' in descriptor) {
+      descriptor.set = undefined;
```
not sure if setting to `undefined` or deleting it makes the most sense

gaearon commented on packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts:
```diff
@@ -59,7 +59,9 @@ function deepCyclicCopyObject<T>(object: T, cycles: WeakMap<any, any>): T {
+    if ('set' in descriptor) {
+      descriptor.set = undefined;
+    } else {
       descriptor.writable = true;
```
This should be outside a condition? You could have a frozen object with a setter on it, right?

SimenB replied:
We cannot set `writable` if there's a getter or a setter. So thanks for the question, we still failed if there was a getter but no setter 🙂 fixed and added a test

```js
Object.defineProperty({}, 'foo', {set(_){}, writable: true}) // Invalid property descriptor. Cannot both specify accessors and a value or writable attribute
Object.defineProperty({}, 'foo', {get(){}, writable: true}) // Invalid property descriptor. Cannot both specify accessors and a value or writable attribute
```

SimenB replied:
Or did you have another example in mind?

SimenB replied:
Added some more tests. With both getter and setter, only getter and only setter, both wrapped in `Object.freeze` and not.

jeysal commented:
I've tried a bit to construct another failure case with a `Proxy`, but we seem to be mostly safe on that side as we will never call its `set`, only `get`, `ownKeys`, `getOwnPropertyDescriptor`, `getPrototypeOf` (all of which things expected to be called on an object being asserted on) and then copy stuff onto something that is inherited from the prototype, but is not itself a Proxy.

gaearon commented on packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts:
```diff
@@ -59,7 +59,9 @@ function deepCyclicCopyObject<T>(object: T, cycles: WeakMap<any, any>): T {
     }
 
-    if (!('set' in descriptor)) {
+    if ('set' in descriptor) {
```
Why can't all of this just be:

```js
const descriptor = descriptors[key];
descriptor = {
  configurable: true,
  enumerable: true,
  writeable: true,
  value: object[key]
}
```

?

gaearon replied:
By the way I don't think `if (typeof descriptor.value !== 'undefined') {` above is right given that `undefined` itself *is* a valid value. The correct check I think is `descriptor.hasOwnProperty('value')`. But regardless, I don't yet see why we can't do it simpler like I wrote above.

jeysal replied:
In theory we probably could, but it makes the assumption that either the getter always returns the same thing, or that the property is only retrieved from the deep copied object once (because in the case of a getter it snapshots the first return value instead of preserving the getter).
The current solution is a more 'accurate' deep copy, although not sure if it matters, because I doubt we support getters with changing return values anyway (the deep equals call and this diff call could already get different values).

SimenB replied:
yeah, that seems to work as well! pushed

gaearon replied:
>In theory we probably could, but it makes the assumption that either the getter always returns the same thing, or that the property is only retrieved from the deep copied object once (because in the case of a getter it snapshots the first return value instead of preserving the getter).

I think in case of printing we _want_ to call the getter as few times as possible, right? Since otherwise the thing we're printing might be different from the thing we were comparing.

That might actually be a good stress test for this whole thing. Have a getter that returns different things on every call. Verify that the comparison shows the diff for the correct result.

jeysal replied:
Yes, I agree that once for every deep property would be ideal. But I think that goal is far away as we already have other get calls happening in a deep equals assertion. Having only one would require the combined deepEqual-prettyPrint-diff implementation (aka the 'proper solution') that we never got around to implementing.
I was viewing this more from a 'deep copy' perspective, in which case preserving getters would be more accurate. This way, it's more of a 'snapshot for diff' method.

jeysal replied:
I don't feel too strongly about this though and think it would be fine to land this way, especially since it's not our standard `deepCyclicCopy` but `deepCyclicCopyReplaceable`.

gaearon replied:
>But I think that goal is far away as we already have other get calls happening in a deep equals assertion.

That seems not great. Is that new behavior in Jest 25, or always been that way?

jeysal replied:
Afaik it has always been like that, because `expect`, `pretty-format`, and `jest-diff` are separate things as opposed to one unified algorithm that generates a 'diff data structure'. `expect` takes the values and checks whether they match (first access to properties). If not, it feeds them into `jest-diff`, which `pretty-format`s both (second access to properties) and then just does a string diff on the two result.
This (as opposed to a unified algorithm for all three tasks) has caused quite some problems esp. regarding asymmetric matchers in the past. I'd love to rewrite the whole thing, but it's huge :sweat_smile:

jeysal replied:
To put the code we're talking about into context, it was introduced recently as a workaround for the asymmetric matchers issues (not the proper unified algorithm fix we'd like) and runs when expect has detected a mismatch, but before the diffing, as a sort of preprocessing.

gaearon replied:
Makes sense.

SimenB commented on packages/jest-matcher-utils/src/Replaceable.ts:
```diff
@@ -36,10 +36,7 @@ export default class Replaceable {
       });
       Object.getOwnPropertySymbols(this.object).forEach(key => {
-        const descriptor = Object.getOwnPropertyDescriptor(this.object, key);
-        if ((descriptor as PropertyDescriptor).enumerable) {
```
everything should be enumerable now

SimenB commented on packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts:
```diff
@@ -49,21 +49,22 @@ export default function deepCyclicCopyReplaceable<T>(
+      configurable: true,
+      enumerable: true,
+      value: deepCyclicCopyReplaceable(
+        (object as Record<string, unknown>)[key],
```
without casting
![image](https://user-images.githubusercontent.com/1404810/78281277-b84ee880-751a-11ea-9de7-d61d4c32b6cf.png)

not sure if it's solvable?

gaearon commented:
Hmm. I must be doing something wrong. I checked out your PR and removed all the implementation changes, but the tests still pass. What am I doing wrong?

SimenB commented:
@gaearon my bad, I simplified the tests so much that they didn't actually fail on master anymore 😬 

Should be fixed now. I also deleted a test that verified getters weren't called (as we explicitly want that now, as per your suggestion)

gaearon commented on packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts:
```diff
@@ -49,21 +49,22 @@ export default function deepCyclicCopyReplaceable<T>(
-    }
-
-    descriptor.configurable = true;
+    descriptors[key] = {
```
Can the enumerable check be here instead?

```js
if (descriptors[key].enumerable) {
  descriptors[key] = ...
} else {
  delete descriptors[key];
}
```

SimenB replied:
in that case we lose `Symbol` keys. Not sure why...

SimenB replied:
At least if I understood your comment correcly


```diff
diff --git i/packages/jest-matcher-utils/src/Replaceable.ts w/packages/jest-matcher-utils/src/Replaceable.ts
index fc8c7343c..dd9e04f32 100644
--- i/packages/jest-matcher-utils/src/Replaceable.ts
+++ w/packages/jest-matcher-utils/src/Replaceable.ts
@@ -35,12 +35,6 @@ export default class Replaceable {
       Object.entries(this.object).forEach(([key, value]) => {
         cb(value, key, this.object);
       });
-      Object.getOwnPropertySymbols(this.object).forEach(key => {
-        const descriptor = Object.getOwnPropertyDescriptor(this.object, key);
-        if (descriptor?.enumerable) {
-          cb(this.object[key], key, this.object);
-        }
-      });
     } else {
       this.object.forEach(cb);
     }
diff --git i/packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts w/packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts
index ad98131dd..67271d5d8 100644
--- i/packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts
+++ w/packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts
@@ -56,15 +56,19 @@ function deepCyclicCopyObject<T>(object: T, cycles: WeakMap<any, any>): T {
   cycles.set(object, newObject);
 
   Object.keys(descriptors).forEach(key => {
-    descriptors[key] = {
-      configurable: true,
-      enumerable: true,
-      value: deepCyclicCopyReplaceable(
-        (object as Record<string, unknown>)[key],
-        cycles,
-      ),
-      writable: true,
-    };
+    if (descriptors[key].enumerable) {
+      descriptors[key] = descriptors[key] = {
+        configurable: true,
+        enumerable: true,
+        value: deepCyclicCopyReplaceable(
+          (object as Record<string, unknown>)[key],
+          cycles,
+        ),
+        writable: true,
+      };
+    } else {
+      delete descriptors[key];
+    }
   });
 
   return Object.defineProperties(newObject, descriptors);
```

gaearon replied:
You'd still need the `Object.getOwnPropertySymbols` call but you can lose the `if (descriptor?.enumerable) {` check because we would've not copied those over in the first place.

SimenB replied:
right! done 🙂

gaearon commented:
I think our longer term aspiration should be that

```js
  let i = 0;
  expect({
    get a() {
      i++;
      return i;
    },
  }).toEqual({ 
    a: 1
   });
```

should pass. But that's out of scope for sure. (I verified it fails on 24.)

SimenB commented:
@jeysal just noticed the discussion you had going in the comments 😅 I'm not able to make the tests pass unless I use the current form of the code. Could you experiment with it? I like copying over the `getter` without invoking it, which the current code does rather than just using `value`

SimenB commented on packages/jest-matcher-utils/src/__tests__/deepCyclicCopyReplaceable.test.ts:
```diff
@@ -20,20 +20,6 @@ test('returns the same value for primitive or function values', () => {
   expect(deepCyclicCopyReplaceable(fn)).toBe(fn);
 });
 
-test('does not execute getters/setters, but copies them', () => {
```
So it's not just in the comments - we invoke the getters on purpose now, so this test doesn't make sense anymore

gaearon commented:
>I like copying over the getter without invoking it,

My concern with this is — are you going to call it on the right object later? What if the getter reads from `this`? Is `this` going to be the real thing or the copy? It needs to be the real thing or you risk breaking other subtle assumptions.

SimenB commented on packages/jest-matcher-utils/src/__tests__/Replaceable.test.ts:
```diff
@@ -104,28 +104,17 @@ describe('Replaceable', () => {
       expect(cb.mock.calls[2]).toEqual([3, symbolKey, object]);
     });
 
-    test('object forEach do not iterate none enumerable symbol key', () => {
```
check moved

SimenB commented:
>My concern with this is — are you going to call it on the right object later?

Yeah, it's called by `pretty-format` when serializing later

gaearon commented:
As long as the getter is called on the original object, seems fine to me. Let's add a test for that?

SimenB commented:
Let's keep it as is now - I don't think copying the getter rather than calling it is observable to the user anyways
